pkg/kubectl/cmd/drain.go

```diff
@@ -195,148 +212,149 @@ func (o *DrainOptions) RunDrain() error {
-	pods, unreplicatedPodNames, daemonSetPodNames, err := GetPodsForDeletionOnNodeDrain(
-		o.client,
-		o.nodeInfo.Name,
-		o.factory.Decoder(true),
-		o.Force,
-		o.IgnoreDaemonsets,
-	)
+func (o *DrainOptions) getController(sr *api.SerializedReference) (interface{}, error) {
+	switch sr.Reference.Kind {
+	case "ReplicationController":
+		return o.client.ReplicationControllers(sr.Reference.Namespace).Get(sr.Reference.Name)
+	case "DaemonSet":
+		return o.client.DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+	case "Job":
+		return o.client.ExtensionsClient.Jobs(sr.Reference.Namespace).Get(sr.Reference.Name)
+	case "ReplicaSet":
+		return o.client.ExtensionsClient.ReplicaSets(sr.Reference.Namespace).Get(sr.Reference.Name)
+	}
+	return nil, fmt.Errorf("Unknown controller kind %q", sr.Reference.Kind)
+}
+
+func (o *DrainOptions) getPodCreator(pod api.Pod) (*api.SerializedReference, error) {
+	creatorRef, found := pod.ObjectMeta.Annotations[controller.CreatedByAnnotation]
+	if !found {
+		return nil, nil
+	}
+
+	// Now verify that the specified creator actually exists.
+	sr := &api.SerializedReference{}
+	if err := runtime.DecodeInto(o.factory.Decoder(true), []byte(creatorRef), sr); err != nil {
+		return nil, err
+	}
+	// We assume the only reason for an error is because the controller is
+	// gone/missing, not for any other cause.  TODO(mml): something more
+	// sophisticated than this
+	_, err := o.getController(sr)
 	if err != nil {
-		return []api.Pod{}, err
+		return nil, err
 	}
+	return sr, nil
+}
 
-	daemonSetErrors := !o.IgnoreDaemonsets && len(daemonSetPodNames) > 0
-	unreplicatedErrors := !o.Force && len(unreplicatedPodNames) > 0
+func (o *DrainOptions) unreplicatedFilter(pod api.Pod) (bool, *warning, *fatal) {
+	sr, err := o.getPodCreator(pod)
+	if err != nil {
+		return false, nil, &fatal{err.Error()}
+	}
+	if sr != nil {
+		return true, nil, nil
+	}
+	if !o.Force {
+		return false, nil, &fatal{kUnmanagedFatal}
+	}
+	return true, &warning{kUnmanagedWarning}, nil
+}
 
-	switch {
-	case daemonSetErrors && unreplicatedErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, daemonSetPodNames, true))
-	case daemonSetErrors && !unreplicatedErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg([]string{}, daemonSetPodNames, true))
-	case unreplicatedErrors && !daemonSetErrors:
-		return []api.Pod{}, errors.New(unmanagedMsg(unreplicatedPodNames, []string{}, true))
+func (o *DrainOptions) daemonsetFilter(pod api.Pod) (bool, *warning, *fatal) {
+	sr, err := o.getPodCreator(pod)
+	if err != nil {
+		return false, nil, &fatal{err.Error()}
+	}
+	if sr == nil || sr.Reference.Kind != "DaemonSet" {
+		return true, nil, nil
+	}
+	if _, err := o.client.DaemonSets(sr.Reference.Namespace).Get(sr.Reference.Name); err != nil {
+		return false, nil, &fatal{err.Error()}
 	}
+	if !o.IgnoreDaemonsets {
+		return false, nil, &fatal{kDaemonsetFatal}
+	}
+	return false, &warning{kDaemonsetWarning}, nil
```

Comment:
No.  No matter what, we never delete the daemonset pods because the DS controller doesn't respect unschedulable.  The flag only toggles whether we consider their presence an error or not.

I will add a comment that clarifies this.